File: LeetCode/Easy/Flood Fill/Go/flood_fill.go
package main

import "fmt"

func flood_fill(arr [][]int, sr, sc int, new_color int) [][]int {
	_flood_fill(&arr, sr, sc, arr[sr][sc], new_color)
	return arr
}

func _flood_fill(arr *[][]int, i, j int, color, new_color int) {
	if i >= len(*arr) || j >= len((*arr)[0]) || i < 0 || j < 0 || (*arr)[i][j] != color {
		return
	}
	(*arr)[i][j] = new_color
	_flood_fill(arr, i-1, j, color, new_color)
	_flood_fill(arr, i, j-1, color, new_color)
	_flood_fill(arr, i+1, j, color, new_color)
	_flood_fill(arr, i, j+1, color, new_color)
}

func main() {
	fmt.Println(flood_fill([][]int{
		{1, 1, 1},
		{1, 1, 0},
		{1, 0, 1},
	}, 1, 1, 2))
}

File: LeetCode/Easy/Flood Fill/Java/floodFill.java
class Solution {
    int[][] dir = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
    boolean[][] visited;
    int n, m;
    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        int c = image[sr][sc];
        n = image.length;
        m = image[0].length;
        visited = new boolean[image.length][image[0].length];
        dfs(image, sr, sc, c, color);
        return image;
    }

    private void dfs(int[][] image, int i, int j, int c, int color) {
        visited[i][j] = true;
        image[i][j] = color;
        for (int k=0; k<4; k++) {
            int x = i+dir[k][0];
            int y = j+dir[k][1];
            if (isSafe(x, y) && !visited[x][y] && image[x][y] == c)
                dfs(image, x, y, c, color);
        }
    }

    private boolean isSafe(int i, int j) {
        return i>=0 && j>=0 && i<n && j<m;
    }
}
